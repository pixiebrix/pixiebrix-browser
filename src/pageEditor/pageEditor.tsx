/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import "@/vendors/bootstrapWithoutRem.css";
import "@/utils/global.scss";
import "@/utils/layout.scss";

import "@/extensionContext";
import "@/development/darkMode.js";

import { initMessengerLogging } from "@/development/messengerLogging";

import ReactDOM from "react-dom";
import React from "react";
import Panel from "@/pageEditor/Panel";
import { watchNavigation } from "@/pageEditor/protocol";
import { initToaster } from "@/utils/notify";
import { initRuntimeLogging } from "@/development/runtimeLogging";
import { initPerformanceMonitoring } from "@/telemetry/performance";
import { setPlatform } from "@/platform/platformContext";
import extensionPagePlatform from "@/extensionPages/extensionPagePlatform";

async function init() {
  setPlatform(extensionPagePlatform);
  void initMessengerLogging();
  void initRuntimeLogging();
  try {
    await initPerformanceMonitoring();
  } catch (error) {
    console.error("Failed to initialize performance monitoring", error);
  }

  watchNavigation();
  initToaster();

  ReactDOM.render(<Panel />, document.querySelector("#container"));
}

void init();
