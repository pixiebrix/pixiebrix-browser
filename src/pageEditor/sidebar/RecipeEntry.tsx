/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { PropsWithChildren } from "react";
import { SemVerString } from "@/core";
import styles from "./Entry.module.scss";
import {
  RecipeHasUpdateIcon,
  UnsavedChangesIcon,
} from "@/pageEditor/sidebar/ExtensionIcons";
import { Accordion, ListGroup } from "react-bootstrap";
import { actions } from "@/pageEditor/slices/editorSlice";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faCaretDown,
  faCaretRight,
  faFile,
} from "@fortawesome/free-solid-svg-icons";
import { useDispatch, useSelector } from "react-redux";
import cx from "classnames";
import {
  selectActiveElement,
  selectDirtyMetadataForRecipeId,
  selectExpandedRecipeId,
  selectRecipeIsDirty,
} from "@/pageEditor/slices/editorSelectors";
import { RecipeDefinition } from "@/types/definitions";
import * as semver from "semver";
import ActionMenu from "@/components/sidebar/ActionMenu";

export type RecipeEntryProps = PropsWithChildren<{
  recipe: RecipeDefinition | undefined;
  isActive?: boolean;
  installedVersion: SemVerString;
  onSave: () => Promise<void>;
  isSaving: boolean;
  onReset: () => Promise<void>;
  onRemove: () => Promise<void>;
}>;

const RecipeEntry: React.FC<RecipeEntryProps> = ({
  recipe,
  isActive,
  children,
  installedVersion,
  onSave,
  isSaving,
  onReset,
  onRemove,
}) => {
  const dispatch = useDispatch();

  const expandedRecipeId = useSelector(selectExpandedRecipeId);
  const activeElement = useSelector(selectActiveElement);
  const {
    id: recipeId,
    name: savedName,
    version: latestRecipeVersion,
  } = recipe?.metadata ?? {};

  // Set the alternate background if an extension in this recipe is active
  const hasRecipeBackground = activeElement?.recipe?.id === recipeId;

  const dirtyName = useSelector(selectDirtyMetadataForRecipeId(recipeId))?.name;
  const name = dirtyName ?? savedName ?? "Loading...";
  const isDirty = useSelector(selectRecipeIsDirty(recipeId));

  const hasUpdate =
    latestRecipeVersion != null &&
    installedVersion != null &&
    semver.gt(latestRecipeVersion, installedVersion);

  const caretIcon = expandedRecipeId === recipeId ? faCaretDown : faCaretRight;

  return (
    <>
      <Accordion.Toggle
        eventKey={recipeId}
        as={ListGroup.Item}
        className={cx(styles.root, "list-group-item-action", {
          [styles.recipeBackground]: hasRecipeBackground,
        })}
        tabIndex={0} // Avoid using `button` because this item includes more buttons #2343
        active={isActive}
        key={`recipe-${recipeId}`}
        onClick={() =>
          recipeId != null && dispatch(actions.selectRecipeId(recipeId))
        }
      >
        <span className={styles.icon}>
          <FontAwesomeIcon icon={faFile} /> <FontAwesomeIcon icon={caretIcon} />
        </span>
        <span className={styles.name}>{name}</span>
        {isDirty && !isActive && (
          <span className={cx(styles.icon, "text-danger")}>
            <UnsavedChangesIcon />
          </span>
        )}
        {hasUpdate && (
          <span className={cx(styles.icon, "text-warning")}>
            <RecipeHasUpdateIcon
              title={`You are editing version ${installedVersion} of this blueprint, the latest version is ${latestRecipeVersion}.`}
            />
          </span>
        )}
        {isActive && (
          <ActionMenu
            onSave={onSave}
            onReset={onReset}
            onRemove={onRemove}
            isDirty={isDirty}
            disabled={isSaving}
          />
        )}
      </Accordion.Toggle>
      <Accordion.Collapse eventKey={recipeId}>
        <>{children}</>
      </Accordion.Collapse>
    </>
  );
};

export default RecipeEntry;
