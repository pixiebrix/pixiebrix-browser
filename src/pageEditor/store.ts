/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore, Middleware } from "@reduxjs/toolkit";
import { persistReducer, persistStore } from "redux-persist";
import { localStorage } from "redux-persist-webextension-storage";
import { editorSlice } from "@/pageEditor/slices/editorSlice";
import { createLogger } from "redux-logger";
import { boolean } from "@/utils";
import { setupListeners } from "@reduxjs/toolkit/query/react";
import { appApi } from "@/services/api";
import runtimeSlice from "@/pageEditor/slices/runtimeSlice";
import { savingExtensionSlice } from "@/pageEditor/panes/save/savingExtensionSlice";
import settingsSlice from "@/store/settingsSlice";
import { persistExtensionOptionsConfig } from "@/store/extensionsStorage";
import servicesSlice, { persistServicesConfig } from "@/store/servicesSlice";
import extensionsSlice from "@/store/extensionsSlice";
import sessionSlice from "@/pageEditor/slices/sessionSlice";
import { logSlice } from "@/components/logViewer/logSlice";
import { authSlice, persistAuthConfig } from "@/auth/authSlice";
import analysisSlice from "@/analysis/analysisSlice";
import pageEditorAnalysisManager from "./analysisManager";
import { tabStateSlice } from "@/pageEditor/tabState/tabStateSlice";
import { recipesSlice } from "@/recipes/recipesSlice";
import { recipesMiddleware } from "@/recipes/recipesListenerMiddleware";

const REDUX_DEV_TOOLS: boolean = boolean(process.env.REDUX_DEV_TOOLS);

const persistSettingsConfig = {
  key: "settings",
  storage: localStorage,
};

const conditionalMiddleware: Middleware[] = [];
if (typeof createLogger === "function") {
  // Allow tree shaking of logger in production
  // https://github.com/LogRocket/redux-logger/issues/6
  conditionalMiddleware.push(
    createLogger({
      // Do not log polling actions (they happen too often)
      predicate: (getState, action) => !action.type.includes("logs/polling"),
    })
  );
}

const store = configureStore({
  reducer: {
    auth: persistReducer(persistAuthConfig, authSlice.reducer),
    options: persistReducer(
      persistExtensionOptionsConfig,
      extensionsSlice.reducer
    ),
    services: persistReducer(persistServicesConfig, servicesSlice.reducer),
    settings: persistReducer(persistSettingsConfig, settingsSlice.reducer),
    editor: editorSlice.reducer,
    session: sessionSlice.reducer,
    savingExtension: savingExtensionSlice.reducer,
    runtime: runtimeSlice.reducer,
    logs: logSlice.reducer,
    analysis: analysisSlice.reducer,
    tabState: tabStateSlice.reducer,
    recipes: recipesSlice.reducer,
    [appApi.reducerPath]: appApi.reducer,
  },
  middleware(getDefaultMiddleware) {
    /* eslint-disable unicorn/prefer-spread -- use .concat for proper type inference */
    return getDefaultMiddleware({
      // See https://github.com/rt2zz/redux-persist/issues/988#issuecomment-654875104
      serializableCheck: {
        ignoredActions: ["persist/PERSIST", "analysis/setKnownVars"],
        ignoredPaths: ["analysis.knownVars"],
      },
      // RTK uses Immer internally, we don't need this extra check
      immutableCheck: false,
    })
      .concat(appApi.middleware)
      .concat(pageEditorAnalysisManager.middleware)
      .concat(recipesMiddleware)
      .concat(conditionalMiddleware);
    /* eslint-enable unicorn/prefer-spread */
  },
  devTools: REDUX_DEV_TOOLS,
});

export const persistor = persistStore(store);

// https://redux-toolkit.js.org/rtk-query/overview#configure-the-store
// Optional, but required for refetchOnFocus/refetchOnReconnect behaviors see `setupListeners` docs - takes an optional
// callback as the 2nd arg for customization
setupListeners(store.dispatch);

export default store;
