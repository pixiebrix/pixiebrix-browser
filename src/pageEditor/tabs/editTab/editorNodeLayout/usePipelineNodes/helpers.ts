/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { DocumentRenderer } from "@/bricks/renderers/document";
import {
  type BrickConfig,
  type BrickPipeline,
  PipelineFlavor,
} from "@/bricks/types";
import {
  type DocumentBuilderElement,
  isButtonElement,
} from "@/pageEditor/documentBuilder/documentBuilderTypes";
import { type SubPipeline } from "@/pageEditor/tabs/editTab/editorNodeLayout/usePipelineNodes/types";
import {
  getDocumentBuilderPipelinePaths,
  getPipelinePropNames,
  getVariableKeyForSubPipeline,
} from "@/pageEditor/utils";
import { type Brick } from "@/types/brickTypes";
import { joinPathParts, joinName } from "@/utils/formUtils";
import { isEmpty, get } from "lodash";

export const ADD_MESSAGE = "Add more bricks with the plus button";

export function getBuilderPreviewElementId(
  brickConfig: BrickConfig,
  path: string,
): string | undefined {
  if (brickConfig.id === DocumentRenderer.BRICK_ID) {
    // The Document Preview element name is a substring of the header node path, e.g.
    // SubPipeline.path: config.body.0.children.9.children.0.children.0.config.onClick.__value__0.children.9.children.0.children.0
    // Document element: 0.children.9.children.0.children.0
    const regex = /config\.body\.(.*)\.config\..*$/;
    const result = regex.exec(path);
    if (result) {
      return result[1];
    }
  }
}

/**
 * @param block the block, or null if the resolved block is not available yet
 * @param blockConfig the block config
 */
export function getSubPipelinesForBlock(
  block: Brick | undefined,
  blockConfig: BrickConfig,
): SubPipeline[] {
  const subPipelines: SubPipeline[] = [];
  if (blockConfig.id === DocumentRenderer.BRICK_ID) {
    for (const docPipelinePath of getDocumentBuilderPipelinePaths(
      blockConfig,
    )) {
      const path = joinPathParts(docPipelinePath, "__value__");
      const pipeline: BrickPipeline = get(blockConfig, path) ?? [];

      // Removing the 'config.<pipelinePropName>' from the end of the docPipelinePath
      const elementPathParts = docPipelinePath.split(".").slice(0, -2);
      const docBuilderElement = get(
        blockConfig,
        elementPathParts,
      ) as DocumentBuilderElement;

      const isButton = isButtonElement(docBuilderElement);

      let subPipelineLabel = docBuilderElement.config.label as string;
      if (isEmpty(subPipelineLabel)) {
        subPipelineLabel = isButton ? "button" : "brick";
      }

      subPipelines.push({
        headerLabel: subPipelineLabel,
        pipeline,
        path,
        flavor: isButton ? PipelineFlavor.NoRenderer : PipelineFlavor.NoEffect,
      });
    }
  } else {
    for (const pipelinePropName of getPipelinePropNames(block, blockConfig)) {
      const path = joinName("config", pipelinePropName, "__value__");
      const pipeline: BrickPipeline = get(blockConfig, path) ?? [];

      const subPipeline: SubPipeline = {
        headerLabel: pipelinePropName,
        pipeline,
        path,
        flavor: PipelineFlavor.NoRenderer,
      };

      const inputKey = getVariableKeyForSubPipeline(
        blockConfig,
        pipelinePropName,
      );

      if (inputKey) {
        subPipeline.inputKey = inputKey;
      }

      subPipelines.push(subPipeline);
    }
  }

  return subPipelines;
}
