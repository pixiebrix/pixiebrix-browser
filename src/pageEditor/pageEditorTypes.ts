/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { AuthRootState } from "@/auth/authTypes";
import { LogRootState } from "@/components/logViewer/logViewerTypes";
import { ExtensionsRootState } from "@/store/extensionsTypes";
import { SavingExtensionState } from "@/pageEditor/panes/save/savingExtensionSlice";
import { SettingsState } from "@/store/settingsTypes";
import { RuntimeState } from "@/pageEditor/slices/runtimeSlice";
import { ExtensionPointType } from "@/extensionPoints/types";
import { RegistryId, UUID } from "@/core";
import { BlockConfig } from "@/blocks/types";
import {
  OptionsDefinition,
  RecipeMetadataFormState,
} from "@/types/definitions";
import { ElementUIState } from "@/pageEditor/uiState/uiStateTypes";
import {
  ActionFormState,
  SidebarFormState,
  TriggerFormState,
  PanelFormState,
  ContextMenuFormState,
  QuickBarFormState,
} from "./extensionPoints/formStateTypes";

export type FormState =
  | ActionFormState
  | SidebarFormState
  | TriggerFormState
  | PanelFormState
  | ContextMenuFormState
  | QuickBarFormState;

export interface EditorState {
  /**
   * A sequence number that changes whenever a new element is selected.
   *
   * Can use as a React component key to trigger a re-render
   */
  selectionSeq: number;

  /**
   * The element type, if the page editor is in "insertion-mode"
   */
  inserting: ExtensionPointType | null;

  /**
   * The uuid of the active element, if an extension is selected
   */
  activeElementId: UUID | null;

  /**
   * The registry id of the active recipe, if a recipe is selected
   */
  activeRecipeId: RegistryId | null;

  /**
   * The registry id of the 'expanded' recipe in the sidebar, if one is expanded
   */
  expandedRecipeId: RegistryId | null;

  error: string | null;

  dirty: Record<string, boolean>;

  /**
   * Unsaved elements
   */
  readonly elements: FormState[];

  /**
   * Brick ids (not UUIDs) that are known to be editable by the current user
   */
  knownEditable: RegistryId[];

  /**
   * True if error is because user does not have access to beta features
   */
  beta?: boolean;

  /**
   * Is the user using the new page editor beta UI?
   */
  isBetaUI: boolean;

  /**
   * The current UI state of each element, indexed by element Id
   */
  elementUIStates: Record<UUID, ElementUIState>;

  /**
   * A clipboard-style-copy of a block ready to paste into an extension
   */
  copiedBlock?: BlockConfig;

  /**
   * Are we currently showing the info message to users about upgrading from v2 to v3 of
   * the runtime api for this extension?
   */
  showV3UpgradeMessageByElement: Record<UUID, boolean>;

  /**
   * Unsaved, changed recipe options
   */
  dirtyRecipeOptionsById: Record<RegistryId, OptionsDefinition>;

  /**
   * Unsaved, changed recipe metadata
   */
  dirtyRecipeMetadataById: Record<RegistryId, RecipeMetadataFormState>;

  // XXX: refactor the is<Modal>Visible state: https://github.com/pixiebrix/pixiebrix-extension/issues/3264

  /**
   * Are we showing the "add extension to blueprint" modal?
   */
  isAddToRecipeModalVisible: boolean;

  /**
   * Are we showing the "remove extension from blueprint" modal?
   */
  isRemoveFromRecipeModalVisible: boolean;

  /**
   * Are we showing the "save as new blueprint" modal?
   */
  isSaveAsNewRecipeModalVisible: boolean;

  /**
   * Are we showing the "create blueprint" modal?
   */
  isCreateRecipeModalVisible: boolean;

  /**
   * When creating a new blueprint from an existing extension, should we keep a separate copy of the extension?
   */
  // XXX: refactor & remove from top-level Redux state. This is a property of the create recipe workflow:
  // https://github.com/pixiebrix/pixiebrix-extension/issues/3264
  keepLocalCopyOnCreateRecipe: boolean;

  /**
   * Unsaved extensions that have been deleted from a recipe
   */
  deletedElementsByRecipeId: Record<RegistryId, FormState[]>;

  /**
   * Newly created recipes that have not been saved yet
   */
  newRecipeIds: RegistryId[];
}

export type RootState = AuthRootState &
  LogRootState &
  ExtensionsRootState & {
    editor: EditorState;
    savingExtension: SavingExtensionState;
    settings: SettingsState;
    runtime: RuntimeState;
  };
