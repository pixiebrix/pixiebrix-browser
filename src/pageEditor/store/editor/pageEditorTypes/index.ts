/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { type AuthRootState } from "@/auth/authTypes";
import { type LogRootState } from "@/components/logViewer/logViewerTypes";
import { type ModComponentsRootState } from "@/store/modComponents/modComponentTypes";
import { type SettingsRootState } from "@/store/settings/settingsTypes";
import { type RuntimeRootState } from "@/pageEditor/store/runtime/runtimeSliceTypes";
import { type UUID } from "@/types/stringTypes";
import { type RegistryId, type VersionedMetadata } from "@/types/registryTypes";
import { type PipelineFlavor } from "@/bricks/types";
import { type AnalysisRootState } from "@/analysis/analysisTypes";
import { type TabStateRootState } from "@/pageEditor/store/tabState/tabStateTypes";
import { type ModDefinitionsRootState } from "@/modDefinitions/modDefinitionsTypes";
import { type SessionChangesRootState } from "@/store/sessionChanges/sessionChangesTypes";
import { type SessionRootState } from "@/pageEditor/store/session/sessionSliceTypes";
import { type ModVariablesDefinition } from "@/types/modDefinitionTypes";
import { type EmptyObject, type Except } from "type-fest";
import { type OptionsArgs } from "@/types/runtimeTypes";
import { type EditorStateMigratedV12 } from "@/pageEditor/store/editor/pageEditorTypes/editorStateMigrated";
import { type ModComponentFormState } from "@/pageEditor/starterBricks/formStateTypes";
import { type EditorStateSynced } from "@/pageEditor/store/editor/pageEditorTypes/editorStateSynced";
import { type EditorStateEphemeral } from "@/pageEditor/store/editor/pageEditorTypes/editorStateEphemeral";

/**
 * Mod-level editor state passed to the runtime/analysis engine
 * @since 2.1.6
 */
export type DraftModState = {
  /**
   * The current option args for the draft mod
   */
  optionsArgs: OptionsArgs;
  /**
   * The current mod variables definition for the draft mod.
   */
  variablesDefinition: ModVariablesDefinition;
};

export type AddBrickLocation = {
  /**
   * The object property path to the pipeline where a block will be added by the add block modal
   */
  path: string;

  /**
   * The flavor of pipeline where a block will be added by the add block modal
   * @see PipelineFlavor
   */
  flavor: PipelineFlavor;

  /**
   * The pipeline index where a block will be added by the add block modal
   */
  index: number;
};

export enum ModalKey {
  MOVE_COPY_TO_MOD,
  SAVE_AS_NEW_MOD,
  CREATE_MOD,
  ADD_BRICK,
  SAVE_DATA_INTEGRITY_ERROR,
}

export type ModalDefinition =
  | { type: ModalKey.ADD_BRICK; data: { addBrickLocation: AddBrickLocation } }
  | { type: ModalKey.SAVE_AS_NEW_MOD; data: EmptyObject }
  | {
      type: ModalKey.CREATE_MOD;
      data: { keepLocalCopy: boolean } & (
        | { sourceModComponentId: UUID }
        | { sourceModId: RegistryId }
      );
    }
  | { type: ModalKey.MOVE_COPY_TO_MOD; data: { keepLocalCopy: boolean } }
  | { type: ModalKey.SAVE_DATA_INTEGRITY_ERROR; data: EmptyObject };

export type ModMetadataFormState = Pick<
  VersionedMetadata,
  "id" | "name" | "version" | "description"
>;

export type EditorState = Except<
  // On migration, re-point this type to the most recent EditorStateV<N> type name
  EditorStateMigratedV12,
  // Swap out any properties with versioned types for type references to the latest version.
  // NOTE: overriding these properties is not changing the type shape/structure. It's just cleaning up the type
  // name/reference which makes types easier to work with for testing migrations.
  "modComponentFormStates"
> & {
  modComponentFormStates: ModComponentFormState[];
} & EditorStateSynced &
  EditorStateEphemeral;

export type EditorRootState = {
  editor: EditorState;
};

export type RootState = AuthRootState &
  LogRootState &
  ModComponentsRootState &
  AnalysisRootState &
  EditorRootState &
  ModDefinitionsRootState &
  TabStateRootState &
  RuntimeRootState &
  SettingsRootState &
  SessionRootState &
  SessionChangesRootState;

export * from "./editorStateEphemeral";
export * from "./editorStateSynced";
export * from "./editorStateMigrated";
