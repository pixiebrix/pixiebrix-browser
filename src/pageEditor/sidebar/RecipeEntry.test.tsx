/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import extensionsSlice from "@/store/extensionsSlice";
import {
  createRenderFunctionWithRedux,
  type RenderFunctionWithRedux,
} from "@/testUtils/testHelpers";
import {
  editorSlice,
  initialState as editorInitialState,
} from "@/pageEditor/slices/editorSlice";
import RecipeEntry, { type RecipeEntryProps } from "./RecipeEntry";
import { type EditorState } from "@/pageEditor/pageEditorTypes";
import { type ExtensionOptionsState } from "@/store/extensionsTypes";
import { validateSemVerString } from "@/types/helpers";
import {
  recipeFactory,
  recipeMetadataFactory,
} from "@/testUtils/factories/recipeFactories";

let renderRecipeEntry: RenderFunctionWithRedux<
  {
    editor: EditorState;
    options: ExtensionOptionsState;
  },
  RecipeEntryProps
>;

beforeEach(() => {
  const recipe = recipeFactory();
  const recipeId = recipe.metadata.id;
  renderRecipeEntry = createRenderFunctionWithRedux({
    reducer: {
      editor: editorSlice.reducer,
      options: extensionsSlice.reducer,
    },
    preloadedState: {
      editor: {
        ...editorInitialState,
        expandedRecipeId: recipeId,
      },
    },
    ComponentUnderTest: RecipeEntry,
    defaultProps: {
      recipe,
      children: <div>test children</div>,
      installedVersion: validateSemVerString("1.0.0"),
      onSave: jest.fn(),
      isSaving: false,
      onReset: jest.fn(),
      onRemove: jest.fn(),
      onClone: jest.fn(),
    },
  });
});

test("it renders", () => {
  const rendered = renderRecipeEntry();

  expect(rendered.asFragment()).toMatchSnapshot();
});

test("renders with empty recipe", () => {
  const rendered = renderRecipeEntry({
    propsOverride: {
      recipe: undefined,
    },
  });

  expect(rendered.asFragment()).toMatchSnapshot();
});

test("renders with empty metadata", () => {
  const recipe = recipeFactory({ metadata: null });
  const rendered = renderRecipeEntry({
    propsOverride: {
      recipe,
    },
  });

  expect(rendered.asFragment()).toMatchSnapshot();
});

test("renders the warning icon when has update", () => {
  const recipe = recipeFactory({
    metadata: recipeMetadataFactory({
      version: validateSemVerString("2.0.0"),
    }),
  });
  const rendered = renderRecipeEntry({
    propsOverride: {
      recipe,
    },
  });

  const warningIcon = rendered.getByTitle(
    "You are editing version 1.0.0 of this mod, the latest version is 2.0.0."
  );

  expect(warningIcon).toBeInTheDocument();
});
