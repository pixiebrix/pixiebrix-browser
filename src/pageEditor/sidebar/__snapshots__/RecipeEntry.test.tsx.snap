// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
<DocumentFragment>
  <div
    class="root list-group-item-action list-group-item"
    tabindex="0"
  >
    <span
      class="icon"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-file fa-w-12 "
        data-icon="file"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M224 136V0H24C10.7 0 0 10.7 0 24v464c0 13.3 10.7 24 24 24h336c13.3 0 24-10.7 24-24V160H248c-13.2 0-24-10.8-24-24zm160-14.1v6.1H256V0h6.1c6.4 0 12.5 2.5 17 7l97.9 98c4.5 4.5 7 10.6 7 16.9z"
          fill="currentColor"
        />
      </svg>
       
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-caret-down fa-w-10 "
        data-icon="caret-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
          fill="currentColor"
        />
      </svg>
    </span>
    <span
      class="name"
    >
      Recipe 1
    </span>
  </div>
  <div
    class="collapse"
  >
    <div>
      test children
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders with empty metadata 1`] = `
<DocumentFragment>
  <div
    class="root list-group-item-action recipeBackground list-group-item"
    tabindex="0"
  >
    <span
      class="icon"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-file fa-w-12 "
        data-icon="file"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M224 136V0H24C10.7 0 0 10.7 0 24v464c0 13.3 10.7 24 24 24h336c13.3 0 24-10.7 24-24V160H248c-13.2 0-24-10.8-24-24zm160-14.1v6.1H256V0h6.1c6.4 0 12.5 2.5 17 7l97.9 98c4.5 4.5 7 10.6 7 16.9z"
          fill="currentColor"
        />
      </svg>
       
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-caret-right fa-w-6 "
        data-icon="caret-right"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 192 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
          fill="currentColor"
        />
      </svg>
    </span>
    <span
      class="name"
    >
      Loading...
    </span>
  </div>
  <div
    class="collapse"
  >
    <div>
      test children
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders with empty recipe 1`] = `
<DocumentFragment>
  <div
    class="root list-group-item-action recipeBackground list-group-item"
    tabindex="0"
  >
    <span
      class="icon"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-file fa-w-12 "
        data-icon="file"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M224 136V0H24C10.7 0 0 10.7 0 24v464c0 13.3 10.7 24 24 24h336c13.3 0 24-10.7 24-24V160H248c-13.2 0-24-10.8-24-24zm160-14.1v6.1H256V0h6.1c6.4 0 12.5 2.5 17 7l97.9 98c4.5 4.5 7 10.6 7 16.9z"
          fill="currentColor"
        />
      </svg>
       
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-caret-right fa-w-6 "
        data-icon="caret-right"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 192 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
          fill="currentColor"
        />
      </svg>
    </span>
    <span
      class="name"
    >
      Loading...
    </span>
  </div>
  <div
    class="collapse"
  >
    <div>
      test children
    </div>
  </div>
</DocumentFragment>
`;
